server/src/main/java/org/elasticsearch/search/aggregations/bucket/significant/heuristics/NXYSignificanceHeuristic.java

```diff
@@ -68,8 +68,19 @@ public void writeTo(StreamOutput out) throws IOException {
     }
 
     @Override
-    public boolean equals(Object other) {
-        return ((NXYSignificanceHeuristic) other).includeNegatives == includeNegatives && ((NXYSignificanceHeuristic) other).backgroundIsSuperset == backgroundIsSuperset;
+    public boolean equals(Object obj) {
+        if (this == obj)
+            return true;
+        if (obj == null)
+            return false;
+        if (getClass() != obj.getClass())
```

Comment:
This changes behaviour, because `NXYSignificanceHeuristic` has subclasses. I cannot myself say whether this matters, so this needs another reviewer.

Reply:
Two objects shouldn't be considered equal if they don't have the same class (their concrete classes could have different state outside of a shared super class).  I would use the IntelliJ generated version, though (shorter):

```
@Override
public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    NXYSignificanceHeuristic that = (NXYSignificanceHeuristic) o;
    return backgroundIsSuperset == that.backgroundIsSuperset &&
        includeNegatives == that.includeNegatives;
}
```

Reply:
Also, @DaveCTurner note that the two classes being the same does not mean they cannot be subclasses.

Reply:
I can adjust that if you prefer?